[6.2] Call events move  to View and add one position for category

korenevskiy opened the pull request:
### Summary of Changes
-Moved the Event call for the Category Blog from the **TMPL** to the **VIEW**. The call has been moved to where it should be initially.
-Added position in the lists of articles of categories, archives selected. This position is displayed after this list of articles. 

This additional event call is necessary for the subsequent withdrawal of information about this list of articles.
It will be convenient to use the list of category articles to create a form of ordering goods. Each material may have a selection field. And at the very end you can place an order form with delivery fields and contact details. And all this will be able to perform one field.

This PR is based on [37789](https://github.com/joomla/joomla-cms/pull/37789) but additionally corrects the placement of event calls in the Categories Blog.


### Testing Instructions
You need to create the `onContentAfterItems($context, $item, $params)` method in any content plugin
Then in this method return additional text.
This text will be inserted into the list of articles of the Blog or other list of articles . 

```PHP
public function onContentAfterItems($context, $item, $params) : string{
	return "<h3>This insert text in list articles</h3>";
}
```
The event of a trigger is caused by this code inserted at the end of the category blog.
```PHP
$results = $app->triggerEvent('onContentAfterItems', array($this->category->extension . '.categories', &$this, &$this->params));
$afterDisplayItems = trim(implode("\n", $results));
echo $afterDisplayItems;
```
### Actual result BEFORE applying this Pull Request
Before all positions allowed to add and process text before to  location position the list of articles.
This approach is convenient for the output of ordinary material, but not for categories and lists of articles .
### Expected result AFTER applying this Pull Request
But when displaying a list of articles, the list itself is important, so you need to display data before the list and after the list of articles.

### Documentation Changes Required
https://docs.joomla.org/Plugin/Events/Content
https://docs.joomla.org/J3.x:Creating_a_content_plugin

korenevskiy committed:
Update blog.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update blog.php

korenevskiy committed:
Update default_articles.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update default.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update default_items.php

korenevskiy committed:
Update default_articles.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update default_items.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Merge branch '4.2-dev' into Move-position-event-to-View-and-Add-one-position-for-category

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

bembelimen commented:
Thank you for your PR. 

There are some challenges with the code:
- Having an old override in place will trigger the events twice, so we can most likely earliest go with Joomla 5 with this
- New events should follow [the new event format](https://github.com/joomla/joomla-cms/blob/4.2-dev/libraries/src/Workflow/WorkflowServiceTrait.php#L97-L108)

korenevskiy commented:
> * New events should follow [the new event format](https://github.com/joomla/joomla-cms/blob/4.2-dev/libraries/src/Workflow/WorkflowServiceTrait.php#L97-L108)

If I redo it to a new format, My PR in J5 will be accepted?
What I need to do to make changes?

korenevskiy commented:
@HLeithner  Hello, in this PR I fixed the CMS Joomla errors that you told me about. please transfer this PR to J5.

HLeithner commented:
@korenevskiy as I already said somewhere I'm not convinced to add events for this kind of template manipulation.
I would like to have something more generic but didn't thought about it yet, I will try to get some people together and see if something more sustainable can be created.

korenevskiy commented:
> as I already said somewhere

You said it here [37789](https://github.com/joomla/joomla-cms/pull/37789) .
 But, I corrected your wishes.
 Before, the call was in the template. I corrected, and now the challenge is in the View.

HLeithner commented:
This pull request has been automatically rebased to 4.3-dev.

korenevskiy commented:
@HLeithner Is it possible to transfer this PR to Joomla 5 ?

Hackwar commented:
Hello @korenevskiy, we discussed this PR in the maintainer meeting today and we don't see the usecase for this. It sounds like a VERY specific circumstance for you. Could you explain why this should be part of Joomla? Why do you need an additional event like this and can't just use a module placed below the content?

korenevskiy commented:
> Hello @korenevskiy, we discussed this PR in the maintainer meeting today and we don't see the usecase for this. It sounds like a VERY specific circumstance for you. Could you explain why this should be part of Joomla? Why do you need an additional event like this and can't just use a module placed below the content?

I've seen a lot of potential in custom fields for articles before. I don't think of creating a single website for a client in space, but I think of creating a plugin for users with a target audience in space.
The plugin must take into account the data fields of each blog article, and then show the final result. For example, blog categories are goods in stock displayed as a list with pictures, then you need to display the total amount and the total number of goods under the list. I also had the idea to create a custom field plugin that would contain a module. It would be very convenient to attach a picture or a table with data to the articles. And then, in the lower position of the category, display a carousel or a basket of goods, data from articles.
But I'm already disappointed to do it. Since this PR and other prs that have a one-line change make great opportunities for developers canceled.

HLeithner commented:
This pull request has been automatically rebased to 5.2-dev.

HLeithner commented:
This pull request has been automatically rebased to 5.3-dev.

richard67 committed:
Merge branch '5.3-dev' into Move-position-event-to-View-and-Add-one-position-for-category

richard67 commented on components/com_content/src/View/Archive/HtmlView.php:
```diff
@@ -133,6 +133,11 @@ public function display($tpl = null)
 
+        $this->event = new \stdClass();
+
+        $results = Factory::getApplication()->triggerEvent('onContentAfterItems', array('com_content.archive', &$this, &$this->params));
+        $this->event->afterDisplayItems = trim(implode("\n", $results));
```
```suggestion
        $results                        = Factory::getApplication()->triggerEvent('onContentAfterItems', array('com_content.archive', &$this, &$this->params));
        $this->event->afterDisplayItems = trim(implode("\n", $results));
```
PHPCS

richard67 commented on components/com_content/src/View/Category/HtmlView.php:
```diff
@@ -179,6 +179,24 @@ public function display($tpl = null)
         }
 
+        $this->category->text = $this->category->description;
+        $app->triggerEvent('onContentPrepare', array($this->category->extension . '.categories', &$this->category, &$this->params, 0));
+        $this->category->description = $this->category->text;
+
+        $this->category->event = new \stdClass();
+
+        $results = $app->triggerEvent('onContentAfterTitle', array($this->category->extension . '.categories', &$this->category, &$this->params, 0));
+        $this->category->event->afterDisplayTitle = trim(implode("\n", $results));
+
+        $results = $app->triggerEvent('onContentBeforeDisplay', array($this->category->extension . '.categories', &$this->category, &$this->params, 0));
+        $this->category->event->beforeDisplayContent = trim(implode("\n", $results));
+
+        $results = $app->triggerEvent('onContentAfterDisplay', array($this->category->extension . '.categories', &$this->category, &$this->params, 0));
+        $this->category->event->afterDisplayContent = trim(implode("\n", $results));
+
+        $results = $app->triggerEvent('onContentAfterItems', array($this->category->extension . '.categories', &$this, &$this->params));
```
```suggestion
        $app->triggerEvent('onContentPrepare',[$this->category->extension . '.categories', &$this->category, &$this->params, 0]);
        $this->category->description = $this->category->text;

        $this->category->event = new \stdClass();

        $results = $app->triggerEvent('onContentAfterTitle', [$this->category->extension . '.categories', &$this->category, &$this->params, 0]);
        $this->category->event->afterDisplayTitle = trim(implode("\n", $results));

        $results = $app->triggerEvent('onContentBeforeDisplay', [$this->category->extension . '.categories', &$this->category, &$this->params, 0]);
        $this->category->event->beforeDisplayContent = trim(implode("\n", $results));

        $results = $app->triggerEvent('onContentAfterDisplay', [$this->category->extension . '.categories', &$this->category, &$this->params, 0]);
        $this->category->event->afterDisplayContent = trim(implode("\n", $results));

        $results = $app->triggerEvent('onContentAfterItems', [$this->category->extension . '.categories', &$this, &$this->params]);
```
PHPCS

richard67 commented on components/com_content/src/View/Featured/HtmlView.php:
```diff
@@ -140,6 +140,11 @@ public function display($tpl = null)
 
+        $this->event = new \stdClass();
+
+        $results = Factory::getApplication()->triggerEvent('onContentAfterItems', array('com_content.featured', &$this, &$this->params));
+        $this->event->afterDisplayItems = trim(implode("\n", $results));
```
```suggestion
        $results                        = Factory::getApplication()->triggerEvent('onContentAfterItems', array('com_content.featured', &$this, &$this->params));
        $this->event->afterDisplayItems = trim(implode("\n", $results));
```

richard67 committed:
Fix PHPCS

richard67 commented on components/com_content/src/View/Featured/HtmlView.php:
```diff
@@ -140,6 +140,11 @@ public function display($tpl = null)
 
+        $this->event = new \stdClass();
+
+        $results                        = Factory::getApplication()->triggerEvent('onContentAfterItems', array('com_content.featured', &$this, &$this->params));
```
```suggestion
        $results                        = Factory::getApplication()->triggerEvent('onContentAfterItems', ['com_content.featured', &$this, &$this->params]);
```

richard67 commented on components/com_content/src/View/Category/HtmlView.php:
```diff
@@ -179,6 +179,24 @@ public function display($tpl = null)
         }
 
+        $this->category->text = $this->category->description;
+        $app->triggerEvent('onContentPrepare',[$this->category->extension . '.categories', &$this->category, &$this->params, 0]);
```
```suggestion
        $app->triggerEvent('onContentPrepare', [$this->category->extension . '.categories', &$this->category, &$this->params, 0]);
```

richard67 commented on components/com_content/src/View/Category/HtmlView.php:
```diff
@@ -179,6 +179,24 @@ public function display($tpl = null)
+
+        $this->category->event = new \stdClass();
+
+        $results = $app->triggerEvent('onContentAfterTitle', [$this->category->extension . '.categories', &$this->category, &$this->params, 0]);
+        $this->category->event->afterDisplayTitle = trim(implode("\n", $results));
+
+        $results = $app->triggerEvent('onContentBeforeDisplay', [$this->category->extension . '.categories', &$this->category, &$this->params, 0]);
+        $this->category->event->beforeDisplayContent = trim(implode("\n", $results));
+
+        $results = $app->triggerEvent('onContentAfterDisplay', [$this->category->extension . '.categories', &$this->category, &$this->params, 0]);
+        $this->category->event->afterDisplayContent = trim(implode("\n", $results));
+
+        $results = $app->triggerEvent('onContentAfterItems', [$this->category->extension . '.categories', &$this, &$this->params]);
+        $this->category->event->afterDisplayItems = trim(implode("\n", $results));
```
```suggestion
        $results                                  = $app->triggerEvent('onContentAfterTitle', [$this->category->extension . '.categories', &$this->category, &$this->params, 0]);
        $this->category->event->afterDisplayTitle = trim(implode("\n", $results));

        $results                                     = $app->triggerEvent('onContentBeforeDisplay', [$this->category->extension . '.categories', &$this->category, &$this->params, 0]);
        $this->category->event->beforeDisplayContent = trim(implode("\n", $results));

        $results                                    = $app->triggerEvent('onContentAfterDisplay', [$this->category->extension . '.categories', &$this->category, &$this->params, 0]);
        $this->category->event->afterDisplayContent = trim(implode("\n", $results));

        $results                                  = $app->triggerEvent('onContentAfterItems', [$this->category->extension . '.categories', &$this, &$this->params]);
        $this->category->event->afterDisplayItems = trim(implode("\n", $results));
```

richard67 committed:
Fix PHPCS (2)

richard67 commented on components/com_content/src/View/Archive/HtmlView.php:
```diff
@@ -133,6 +133,11 @@ public function display($tpl = null)
 
+        $this->event = new \stdClass();
+
+        $results                        = Factory::getApplication()->triggerEvent('onContentAfterItems', array('com_content.archive', &$this, &$this->params));
```
```suggestion
        $results                        = Factory::getApplication()->triggerEvent('onContentAfterItems', ['com_content.archive', &$this, &$this->params]);
```

richard67 committed:
Fix PHPCS (3)

HLeithner commented:
This pull request has been automatically rebased to 6.0-dev.

korenevskiy committed:
Merge branch '6.0-dev' into Move-position-event-to-View-and-Add-one-position-for-category

brianteeman commented:
> Hello @korenevskiy, we discussed this PR in the maintainer meeting today and we don't see the usecase for this. It sounds like a VERY specific circumstance for you. Could you explain why this should be part of Joomla? Why do you need an additional event like this and can't just use a module placed below the content?

Now that maintainers have a reply can you please make a decision. There is no point in leaving it open forever and continually rebasing the branch.

korenevskiy commented:
> > Hello @korenevskiy, we discussed this PR in the maintainer meeting today and we don't see the usecase for this. It sounds like a VERY specific circumstance for you. Could you explain why this should be part of Joomla? Why do you need an additional event like this and can't just use a module placed below the content?
> 
> Now that maintainers have a reply can you please make a decision. There is no point in leaving it open forever and continually rebasing the branch.

Well, first of all, I cleaned up the code and moved the plugin call from the template to the view. This is a big part of the fix.  By adding a generic plugin method call. I could use a module. 

But I was writing a plugin that does not solve the problem of one customer, I was writing a plugin for sale in general for a large number of customers. Customers should have no problems when they install the plugin themselves. That is, I made a plugin that performs the function of a store, i.e. the entire store in one plugin. Of course, this is my personal problem. 

BUT! BUT! BUT!. This allowed me to understand that the site materials can be used as a product, without components, without copying data, without converting data. publications that are already on the site become an order item. With the sales components, you need to synchronize the product description with the article description. In the case when articles are like a product: you can use modules: popular, recent, sliders. There is no need to look for something special when existing modules work with articles. And you don't need to redo the site at all. You don't need to customize new components to fit the site's style. Everything has already been checked, we need the order form or the shopping cart form to be displayed at the bottom of the page. BUT if you use articles only as articles, then of course my idea is too intrusive. And of course, my request is easier to solve on the module. If I were doing this for only one client, I would solve the problem only using the module. Using the module, everyone does this, they insert entire components into the module, the module of which is inserted into the middle of the article. Frequent occurrence. But I was making a free plugin for everyone.
allowing people not to worry about the site in principle. By installing the module. 
**PlugInPlay**, In the case of a new store component, this will not work.

HLeithner commented:
This pull request has been automatically rebased to 6.1-dev.

korenevskiy committed:
Create ItemsDisplayEvent.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Merge branch '6.2-dev' into Move-position-event-to-View-and-Add-one-position-for-category

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update default_items.php

korenevskiy committed:
Update blog.php

korenevskiy committed:
Update default_articles.php

korenevskiy committed:
Update default.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update ItemsDisplayEvent.php

korenevskiy committed:
Update ItemsDisplayEvent.php

korenevskiy committed:
Update ItemsDisplayEvent.php

QuyTon commented on components/com_content/src/View/Archive/HtmlView.php:
```diff
@@ -81,6 +82,13 @@ class HtmlView extends BaseHtmlView
+    /**
+     * Prepared dispatcher result for category
+     * @var  \stdClass
+     * @since  6.2
```
```suggestion
     * @since  __DEPLOY_VERSION__
```
Please repeat where applicable.

korenevskiy committed:
Update components/com_content/src/View/Archive/HtmlView.php

Co-authored-by: Quy Ton <quy@nomonkeybiz.com>

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update ItemsDisplayEvent.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update HtmlView.php

korenevskiy committed:
Update ItemsDisplayEvent.php

korenevskiy committed:
Update ItemsDisplayEvent.php
